src/annotator/draw-tool.tsx

```diff
@@ -0,0 +1,230 @@
+
+  /** Controller for ending the drawing operation. */
+  private _abortDraw?: AbortController;
+
+  /**
+   * @param root - Container in which the user can draw a shape
+   */
+  constructor(root: HTMLElement) {
+    this._container = root;
+    this._tool = 'rect';
+    if (!isPositioned(root)) {
+      root.style.position = 'relative';
+    }
+  }
+
+  destroy() {
+    this.cancel();
+  }
+
+  /**
+   * Begin drawing a shape.
+   *
+   * @param tool - Type of shape to draw
+   * @return - Promise for the shape drawn by the user
+   */
+  async draw(tool: Tool): Promise<Shape> {
+    this._tool = tool;
+
+    // Only one drawing operation can be in progress at a time.
+    this.cancel();
+
+    // Create a transparent SVG canvas overlaid on top of the container, with
+    // a crosshair cursor to indicate the user can click to draw.
+    const surface = document.createElementNS(
+      'http://www.w3.org/2000/svg',
+      'svg',
+    );
+    surface.setAttribute('data-testid', 'surface');
+    surface.setAttribute('width', '100%');
+    surface.setAttribute('height', '100%');
+    surface.style.cursor = 'crosshair';
+    surface.style.position = 'absolute';
+    surface.style.left = '0px';
+    surface.style.top = '0px';
+    surface.style.zIndex = '10';
+    this._container.append(surface);
+    this._surface = surface;
+
+    let resolve: (s: Shape) => void;
+
+    const shape = new Promise<Shape>((resolve_, reject_) => {
+      resolve = resolve_;
+      this._drawEnd = resolve_;
+      this._drawError = reject_;
+    });
+
+    this._surface.addEventListener('mousedown', e => {
+      switch (this._tool) {
+        case 'rect':
+          this._shape = {
+            type: 'rect',
+            left: e.clientX,
+            top: e.clientY,
+            right: e.clientX,
+            bottom: e.clientY,
+          };
+          this._renderSurface();
+          break;
+        case 'point':
+          resolve({
+            type: 'point',
+            x: e.clientX,
+            y: e.clientY,
+          });
+          this._abortDraw?.abort();
+          break;
+      }
+    });
+
+    this._surface.addEventListener('mousemove', e => {
+      if (!this._shape) {
+        return;
+      }
+
+      this._shape.right = e.clientX;
+      this._shape.bottom = e.clientY;
+      this._renderSurface();
+    });
+
+    this._surface.addEventListener('mouseup', e => {
+      if (!this._shape) {
+        return;
+      }
+      this._shape.right = e.clientX;
+      this._shape.bottom = e.clientY;
+      resolve(normalizeRect(this._shape));
+      this._abortDraw?.abort();
+    });
+
+    // Cancel drawing if user presses "Escape"
+    this._abortDraw = new AbortController();
+    document.body.addEventListener(
+      'keydown',
+      (e: KeyboardEvent) => {
+        if (e.key === 'Escape') {
+          if (this._drawError) {
+            this._drawError(new DrawError('Drawing canceled'));
+          }
+          this._abortDraw?.abort();
+        }
```

Comment:
Perhaps we can avoid too much nesting with an early return.

```suggestion
        if (e.key !== 'Escape') {
          return;
        }
        
        if (this._drawError) {
          this._drawError(new DrawError('Drawing canceled'));
        }
        this._abortDraw?.abort();
```